### ISSUE TITLE
LFR Generator fails randomly

### ISSUE THREAD

**Issue by michitux** [kit-parco/networkit#171]
The LFR generator draws random community sizes from a certain distribution (with the given parameters) and draws random node degrees from a certain distribution (again with given parameters). It then tries to assign nodes to communities. Each node u needs to have (1 - mu) * deg(u) neighbors in its own community. Depending on the random community sizes drawn and the random node degrees drawn and depending on the given parameters, it might be that this assignment is just completely impossible as e.g. the largest community is just too small to fit the node of the highest degree. How likely this is totally depends on the given parameters but it definitely happens relatively frequently with parameters reported in the literature.

The original implementation just merges communities if this happens until all nodes fit. Imho this is very bad as if you choose your parameters unwisely, you end up with completely different community sizes. That's why I did not implement this but just let the generator fail.

Unfortunately, this means that in code where you use the LFR generator, you might need to write loops to try calling the generator several times until it succeeds. I think it would be nice if this was not needed as, as discussed in #170, this is not so nice from a user's perspective.

The code has some sanity checks to test if finding a fitting distribution is plain impossible and fails earlier then. However, I think it is quite possible to supply parameters where finding a fitting distribution is highly unlikely. Therefore, I am not sure if just re-drawing distributions until a fitting distribution is drawn is a good idea. We might of course add more sanity checks to detect such edge cases. Another possibility would be to only re-draw a certain number of times, which would simply make a failure more unlikely (but it would of course still be possible). Due to the (very flexible) interface of the generator I'm also not sure where exactly the re-drawing should happen. However, a different, less flexible but more convenient interface could also be added while properly fixing #26.

Do you have any other suggestions what the desired behavior should be?

**Comment by avdgrinten** [comment]
Is it possible to detect cases in which the generator fails with high probability (e.g. > 50% or whatever)? In that case we could just fail on those cases and loop on the others. However, given your description, that seems hard to do, so we might need a more complex solution.

**Comment by michitux** [comment]
I don't think it is easily possible to determine the failure probability. If it is likely to have a community in which the node of maximum degree fits, the generator should succeed in most cases. However, it might also succeed if having a node of such a high degree is not likely. Apart from that it might of course also fail if there are too many small communities but not enough nodes with small degree. All of this of course also depends on the exponents of the power law distributions.

I would suggest to have a parameter for the number of retries and set it by default e.g. to 10. I would assume that this should make failure really unlikely for frequently used parameter sets but it would still prevent that the algorithm runs forever. This would require some refactoring though as I think this is not possible with the current interface and data structure.
